source/grid/grid_generator.cc

```diff
@@ -1910,6 +1910,313 @@ namespace GridGenerator
+                                                        0.41 / 4.0);
+        GridTools::shift(cylinder_triangulation_offset, cylinder_tria);
+        // Mark the cylinder material ids as 2 to distinguish them from the bulk
+        // cells (these are read again about 40 lines down)
+        for (const auto &cell : cylinder_tria.active_cell_iterators())
+          cell->set_material_id(2);
+
+        auto minimal_line_length = [](const Triangulation<2> &t) -> double {
+          double min_line_length = std::numeric_limits<double>::max();
+          for (const auto &cell : t.active_cell_iterators())
+            for (unsigned int line_n = 0;
+                 line_n < GeometryInfo<2>::lines_per_cell;
+                 ++line_n)
+              min_line_length = std::min(min_line_length,
+                                         (cell->line(line_n)->vertex(0) -
+                                          cell->line(line_n)->vertex(1))
+                                           .norm());
+          return min_line_length;
+        };
+
+        // the cylindrical triangulation might not match the Cartesian grid: as
+        // a result the vertices might not be lined up. Get around this by
+        // deleting duplicated vertices with a very low numerical tolerance.
+        const double tolerance =
+          std::min(minimal_line_length(tria_without_cylinder),
+                   minimal_line_length(cylinder_tria)) /
+          2.0;
+
+
+        GridGenerator::merge_triangulations(tria_without_cylinder,
+                                            cylinder_tria,
+                                            tria,
+                                            tolerance);
+
+        const types::manifold_id polar_manifold_id = 0;
+        const types::manifold_id tfi_manifold_id   = 1;
+        for (const auto &cell : tria.active_cell_iterators())
+          {
+            // set all non-boundary manifold ids on the cells that came from the
+            // grid around the cylinder to the new TFI manifold id.
+            if (cell->material_id() == 2)
+              {
+                cell->set_manifold_id(tfi_manifold_id);
+                for (unsigned int face_n = 0;
+                     face_n < GeometryInfo<2>::faces_per_cell;
+                     ++face_n)
+                  {
+                    if (cell->face(face_n)->at_boundary())
+                      cell->face(face_n)->set_manifold_id(polar_manifold_id);
+                    else
+                      cell->face(face_n)->set_manifold_id(tfi_manifold_id);
+                  }
+              }
+          }
+        // ensure that all other manifold ids (including the faces opposite
+        // the cylinder) are set to the flat id
+        for (const auto &cell : tria.active_cell_iterators())
+          if (cell->material_id() != 2)
+            cell->set_all_manifold_ids(numbers::flat_manifold_id);
+
+        // We need to calculate the current center so that we can move it later:
```

Comment:
maybe move the whole center shift part to `hyper_cube_with_cylindrical_hole` via additional parameter? Can be useful elsewhere.